쥬스 메이커 [STEP 2] Doku, DonggeonOh

donggeonoh opened the pull request:
담당 리뷰어: @GREENOVER 😘
범위: STEP 1 에서 Model의 약간의 수정사항 및 STEP 2 전부

### 요구사항 별 구현 방법
1. 재고수정 버튼을 터치 시 재고 추가 화면으로 이동
재고 수정 버튼을 터치 시 구현은 STEP 3 에서 마무리 할 예정입니다.
현재 재고 수정 버튼을 눌렀을 때 NavigationViewController를 통한
재고 추가 화면 이동에 대한 기능 구현은 완료 되었으나
STEP 3 와 현재 동시 진행을 하고 있어서 Merge conflict 가 우려되어
재고 추가 화면에 storyboardID 값을 지정하지 않았습니다.

2. 주문 버튼(셀) 구현
주문 버튼 구현은 Button 을 통하여 구현 하지 않았고 UICollectionView를 이용하여 Cell을 터치하여 구현하였습니다.
Juice 열거형 타입의 케이스가 줄어들거나 늘어나는 것에 대해 동적으로 주문 버튼 셀을 보여주기 위하여
UICollectionView를 이용하였습니다.

3. 주문 버튼을 터치 시
  -  공통
  JuiceMaker의 make 메서드를 호출하면 StoreError를 통해 Error를 반환 받을 수 있기 때문에
  error handling을 하여 각각의 상황에 맞게 AlertController를 띄우게 하였습니다.
  그리고 Alert을 띄우기 위해 성공 했을 때의 Alert과, 실패 했을 때의 Alert의 구현을
  extension을 통해 구현 하였습니다.
  - 재고가 있는 경우
  Alert이 표시되기 전 JuiceMaker의 make 메서드를 호출 하였으므로 FruitStoreManager에서 재고가 반영됩니다.
  UIAlertController.makeSuceesJuiceAlert(juice: Juice)를 호출하여
  화면에 어떤 쥬스를 만들었는지 Alert 창을 띄웁니다.
  - 재고가 없는 경우
    - 재고가 0인 경우
      UIAlertController.makeFailJuiceAlert(juice: Juice)를 호출하여  어떤 재료가 없는지 Alert을 띄워줍니다.
    - 재고가 부족한 경우
      UIAlertController.makeFailJuiceAlert(juice: Juice, count: Int)를 호출하여 어떤 재료가 얼마만큼 남았는지 Alert을 띄워줍니다. 

### 고민됐던 점
1. 문자열, ID, 상수 관리의 문제 및 궁금한 점
현재 상수에 관련된 값을 NameSpace.swift에 저장 하였습니다.
근데 View의 ID, Notification.Name, Alert 이나 log에 필요한 String 값, 기본으로 재고를 추가하기 위한 [Fruit:Int] 타입의 initValue, Layout과 관련된 상수를 한군데에 몰아넣다 보니 점점 복잡해졌습니다.
그래서 이런 값들을 관리하는 다수의 타입을 갖는 것이 유리한지 그리고 enum 안에 static let으로 값을 묶어 놔서 관리하고 있는데 외에 다른 효율적으로 상수 값을 접근 할 수 있는 방법이 있을까요?

2. 2개의 UICollectionView를 한 UIViewController에 정의하여 View controller가 거대해지는 문제
과일의 재고를 보여주는 컬렉션뷰와 주문 버튼에 대한 컬렉션 뷰를 하나의 뷰컨트롤러에서 관리 중입니다.
delegate나 datasource 같은 곳에서 switch 문을 통한 각각의 컬렉션 뷰의 코드를 작성하였습니다.
이전 그린이 메서드를 분리하라고 말씀 하셨지만, 만약에 컬렉션뷰 안에 컬렉션뷰가 있거나
컬렉션뷰를 과장하여 10개를 가지는 복잡한 경우에는 어떻게 관리해야 할까요?

3. collectionView.reloadData()에 대한 오버헤드 관련 궁금한 점 및 특정 셀에 대한 접근방법
쥬스를 만들면(만드는 것을 성공하는 경우) 재고를 Model에 반영하게 되는데,
collectionView.reloadData() 호출에 있어서 오버헤드가 커지는 것에 대한 문제는 현재 없지만
만약 1000개 그 이상이거나 Cell 자체에 가지고 있는 뷰가 복잡한 경우엔 오버헤드가 커질 수 있다고 생각합니다.
이런 경우엔 특정 셀에 대한 접근을 어떻게 하는 것이 좋을까요?

DOAAANG committed:
feat: Add method to update fruit label text

donggeonoh committed:
feat: Add access name and emoji by offest into Fruit

DOAAANG committed:
feat: Implement order button
음료 주문 버튼시 FruitStore 재고에 따라 얼럿 컨트롤러 나타나개 구현

DOAAANG committed:
feat: Add show modify stock view controller action

Right navigation bar item을 통한 뷰 전환
재고 부족시 얼럿 액션을 통한 뷰 전환

DOAAANG committed:
style: Add Name space for storyboard alert action

DOAAANG committed:
refactor: Refactor Fruit stock label Order button to Collection view
기존Label과 Button으로 구성한 View를 재사용과 확작성을 고려하여 CollectionView로 구현

DOAAANG committed:
chore: Add NameSpace.swift for move constants enum
View Controller에서 작성한 StoryboardID StoryboardName AlertTitle 생성한 NameSpace파일로 이동

DOAAANG committed:
feat: Add NotificationCenter to receive data
View Controller 간 객체간 소통을 위하여 Callback, KVO, Delegate, Notification 중 고민하다 Notification 채택하여 적용

DOAAANG committed:
Merge branch 'feat/step-2-juice-maker-implent-inital-view' into feat/step-2-inital-view

DOAAANG committed:
feat: Change fruit store type FruitStore to FruitStorable

fruitStore과 juiceMaker를 optional로 선언함으로 초기화하는 메소드 추가
오토레이아웃 수정

DOAAANG committed:
style: Move method at top

donggeonoh committed:
refactor: Separate more then 10 line of methods in InitialVC and Gather contant in Enum
InitialVC의 메서드들에서 10줄을 초과하면 메서드를 분리
AlertController의 경우 팩토리 패턴으로 Alert 클래스를 만들어주는 팩토리 클래스 생성
메서드안에 있는 상수를 enum 타입으로 정리

donggeonoh committed:
docs: Add Markup comment in InitialViewController

donggeonoh committed:
docs: Add Daily scrum and work history in README.md

donggeonoh committed:
docs: Change grammer in README.md

GREENOVER commented on JuiceMaker/JuiceMaker/Controller/InitialViewController/InitialViewController.swift:
```diff
@@ -0,0 +1,181 @@
+    }
+    
+    private func showCompleteMakeAlert(about juice: String) {
+        let alert = UIAlertController.makeSuceesJuiceAlert(juice: juice)
```
UIViewController를 익스텐션해서 얼럿을 만들면 static으로 해줄 필요가 없이 더 간결하게 메서드 호출할 수 있을것 같은데 어떤 의견이신가요?🤔

GREENOVER replied:
UIViewController를 말했습니다!
Alert 코드를 분리하지 않고 해당 VC에서 생성해서 사용해도 전혀 상관없고 무방합니다~!
고민해보시면 될거같아요ㅎㅎ

GREENOVER commented on JuiceMaker/JuiceMaker/Extension/UIAlertController+Extension.swift:
```diff
@@ -0,0 +1,42 @@
+import UIKit
+
+extension UIAlertController {
+    static func makeSuceesJuiceAlert(juice: String) -> UIAlertController {
```
UIVC을 익스텐션하면 static을 해줄 이유가 없다고 생각합니다🤔

DonggeonOh replied:
위와 내용 동일합니다!

GREENOVER commented on JuiceMaker/JuiceMaker/View/FruitStockCollectionViewCell.swift:
```diff
@@ -0,0 +1,18 @@
+import UIKit
+
+class FruitStockCollectionViewCell: UICollectionViewCell {
+    @IBOutlet weak var fruitEmojiLabel: UILabel!
```
저는 이모지보다 이미지로 사용해야하지 않나 생각합니다🤔
Fruit의 emoji 프로퍼티가 여기에 활용되는거였군요..!
그런데 사실 요걸 이모지 스트링으로 하면 문제가 없을까요?
추후 과일들의 이미지가 이모지에 없는걸로 교체되면 코드 전체를 리팩토링 해야하지 않을까요..?

GREENOVER commented on JuiceMaker/JuiceMaker/Controller/InitialViewController/InitialViewController.swift:
```diff
@@ -0,0 +1,181 @@
+    }
+    
+    // MARK: - Initialization
+    func initConfigrations() {
```
또한 이 메서드는 외부에서 호출되나요~?

iDoyoung replied:
아 수정하겠습니다!

iDoyoung replied:
조금 더 네이밍에 고민해 보겠습니다

GREENOVER commented on JuiceMaker/JuiceMaker/Extension/UIAlertController+Extension.swift:
```diff
@@ -0,0 +1,42 @@
+        return alert
+    }
+    
+    private static func formatFaildTitle(juice: String, count: Int?) -> String {
```
titleFormatFailed 혹은 FailedTitleFormat이 좀 더 어순이 맞는 네이밍이 아닐까 생각합니다🤔

GREENOVER reviewed (changes requested):
@DonggeonOh @iDoyoung 
스텝 2 고생 많으셨습니다!
음.. 우선 리뷰에 앞서 다시 한번 스텝2를 점검해보셔야할것 같아요ㅠ
가장 큰 문제는 2가지입니다.
1. 재고수정 화면으로 이동 시 앱 크래쉬 발생
2. 뷰의 구현이 요구사항과 다름
![Simulator Screen Shot - iPhone 11 - 2022-02-18 at 09 37 35](https://user-images.githubusercontent.com/72292617/154595694-e03a8ea0-9bf7-490d-953b-7e3a1e2ca3e7.png)
주문 버튼이 요구사항과 상당히 다릅니다.
컬렉션뷰를 사용해서 이후 더 편리한 관리와 틀을 만드신건 이해가 됩니다!
그러나 요구사항은 개발자와 디자이너/기획과의 회의 및 약속으로 인해 도출된 경우가 많아요.
가장 중요한건 개발자는 이 요구사항을 지켜야됩니다.
만약 요구사항이 납득이 안가거나 다른 요구사항이 더 좋을것 같은데? 라는 의견이 있으면 꼭 디자이너와 기획과 상의하고 받아들여지고 진행되야합니다!
이건 정말 정말 중요해요..🥲
또한 현재 구현의 디자인으로는 스크롤을 해서 쥬스 주문 버튼을 찾는등의 액션도 취해져야하는데 사용자의 입장으로도 상당히 불편합니다ㅠㅠ
그 외에 우선 수정 후 재리뷰를 하기전 꼭 필요한 변경이나 보완해야할 포인트들을 코멘트로 남겨봤습니다!🙌
***
### 고민하신 부분에 대한 저의 의견
1. 문자열, ID, 상수 관리의 문제 및 궁금한 점
- 저는 네임스페이스 관리 괜찮다고 생각해요!
코멘트에도 남겼지만 조금 더 파일 구조만 가져가면 좋을것 같습니다!
2. 2개의 UICollectionView를 한 UIViewController에 정의하여 View controller가 거대해지는 문제
- 이 부분에 있어 제가 아마 메서드로 분리하는것도 말씀드리면서 하나의 delegate에 다수의 뷰를 연결지을 수 있다고 말씀드렸던 기억이 있습니다..!
그 키워드로 찾아보시면 어떨까요? 아주 많은 레퍼런스들이 나올거에요😀
3. collectionView.reloadData()에 대한 오버헤드 관련 궁금한 점 및 특정 셀에 대한 접근방법
- 어떤 오버헤드를 말씀하시는걸까요?
요소가 많아지면 리로드데이터가 되면 다시 그려지는데 버벅이거나 하는것을 말하는걸까요?
만약 그런거라면 추후 페이지네이션을 이용해볼 수 있을것 같은데,
사실 통신을해서 이미지를 받아와 컬렉션뷰에 뿌려주고 이런 과정들이 아닌 이미 메모리에 들어있는것들로 다뤄주기에 오버헤드가 날정도인가?
라는 생각은 듭니다!
***
수고많으셨어요!
위에서 언급했듯이 무조건 수정이 이뤄져야하는 사항들이 있다보니 request Changes하겠습니다!
수정되면 리뷰 재요청 부탁드려요🙌

iDoyoung commented on JuiceMaker/JuiceMaker/Constant/NameSpace.swift:
```diff
@@ -0,0 +1,70 @@
+//
+//  NameSpace.swift
```
그렇게 하느것이 파악하는데 있어서 좋겠네요. 감사합니다!

iDoyoung commented on JuiceMaker/JuiceMaker.xcodeproj/project.pbxproj:
```diff
@@ -9,15 +9,20 @@
 		C73DAF37255D0CDD00020D38 /* AppDelegate.swift in Sources */ = {isa = PBXBuildFile; fileRef = C73DAF36255D0CDD00020D38 /* AppDelegate.swift */; };
 		C73DAF39255D0CDD00020D38 /* SceneDelegate.swift in Sources */ = {isa = PBXBuildFile; fileRef = C73DAF38255D0CDD00020D38 /* SceneDelegate.swift */; };
-		C73DAF3B255D0CDD00020D38 /* ViewController.swift in Sources */ = {isa = PBXBuildFile; fileRef = C73DAF3A255D0CDD00020D38 /* ViewController.swift */; };
+		C73DAF3B255D0CDD00020D38 /* InitialViewController.swift in Sources */ = {isa = PBXBuildFile; fileRef = C73DAF3A255D0CDD00020D38 /* InitialViewController.swift */; };
```
아 그러면 JuiuceOrederViewContriller 로 변경하도록 하겠습니다

DonggeonOh replied:
Doku 오타가 났네용 JuiceOrderViewController 맞죠?

DonggeonOh commented on JuiceMaker/JuiceMaker/Constant/NameSpace.swift:
```diff
@@ -0,0 +1,70 @@
+enum AlertTitle {
+    static let completeMakeJuice = "%@가 나왔습니다! 맛있게 드세요!"
+    static let notEnoughJuice = "%@의 갯수가 %d개 밖에 없어요 재고를 수정할까요?"
+    static let faliMakeJuice = "%@가 없어요. 재고를 수정할까요?"
```
네 맞습니다~ 재고가 아예 0 이거나 재고는 있지만 부족한 경우 StoreError에 구분 지어서 케이스를  분리하였습니다.
Error에 대한 케이스를 나누려고 했는데 나눌게 몇개 없어서 좀 더 구체적으로 나눠봤습니다.. ㅎㅎ 연습할겸 말이죵.. 사실상 notEnoughStock error 하나로 합쳐 구현가능합니다. 네이밍에도 혼동이 와서 네이밍도 바꿔보겠습니다.

DonggeonOh commented on JuiceMaker/JuiceMaker/Controller/InitialViewController/InitialViewController.swift:
```diff
@@ -0,0 +1,181 @@
+        present(alert, animated: true)
+    }
+    
+    private func showFailedMakeAlert(about juice: String, count: Int? = nil) {
```
네이밍에 관한 조언 항상 감사드립니다. 둘이 고민하면 답이 안나와서 참 이게 문제네요 ㅠㅠ 네이밍에 관한 코멘트는 정말 언제나 환영입니다 ㅎㅎ..

DonggeonOh replied:
와 정말 이런 꿀팁 진짜 꿀입니다.. 감사합니다!!! Alamofire와 같이 많이 사용되는 라이브러리를 보면서 네이밍을 어떻게 하는지 봐야겠어요 ㅎㅎㅎ

DonggeonOh commented on JuiceMaker/JuiceMaker/Extension/FruitStockManager+Extension.swift:
```diff
@@ -0,0 +1,12 @@
+//  Created by Oh Donggeon on 2022/02/17.
+//
+
+import Foundation
+
+extension FruitStockManager {
+    
+}
```
아뇽 ㅠㅠ 이부분 삭제하겠습니다~

DonggeonOh commented on JuiceMaker/JuiceMaker/Extension/NotificationCenter+Extension.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+extension Notification.Name {
+    static let sendFruitStoreData = Notification.Name(rawValue: "SendFruitData")
+}
```
그렇다면 익스텐션 보다는 상수로 관리하는 것이 맞는 것 같네요~ 저렇게 사용하려던 의도는 Notification 에서 abbOberver 시 Name 매개변수를 입력 할 때 . 키워드로 바로 넘겨주려고 이런 방식으로 구현했네용 ㅎㅎ.. (그리고 앞에 타입이 또 붙어버리면 너무 이름이 길어지는 부분이 있어서요)
하지만 그린이 말씀하신 extension의 의미와 달라 빼는 게 나아 보입니다.

DonggeonOh commented on JuiceMaker/JuiceMaker/Extension/UIAlertController+Extension.swift:
```diff
@@ -0,0 +1,42 @@
+//  Created by Oh Donggeon on 2022/02/17.
+//
+
+import Foundation
```
앗 감사합니다 그 부분 알고 있었는데 까먹고 있었네용 ㅎㅎ..

DonggeonOh commented on JuiceMaker/JuiceMaker/Extension/UIAlertController+Extension.swift:
```diff
@@ -0,0 +1,42 @@
+import UIKit
+
+extension UIAlertController {
+    static func makeSuceesJuiceAlert(juice: String) -> UIAlertController {
```
네이밍 통일하겠습니다~

DonggeonOh commented on JuiceMaker/JuiceMaker/Extension/UIAlertController+Extension.swift:
```diff
@@ -0,0 +1,42 @@
+import UIKit
+
+extension UIAlertController {
+    static func makeSuceesJuiceAlert(juice: String) -> UIAlertController {
+        let alert = UIAlertController(title: String(format: AlertTitle.completeMakeJuice, juice),
+                                      message: nil,
+                                      preferredStyle: .alert)
+        let okayAlertAction = UIAlertAction(title: AlertTitle.action.submit,
+                                            style: .default)
+        alert.addAction(okayAlertAction)
+        return alert
+    }
+    
+    static func makeFailJuiceAlert(juice: String, count: Int?, completion: @escaping (UIAlertAction) -> Void) -> UIAlertController {
```
익스텐션 말고 차라리 UIAlertController를 필요 할때 마다 생성해서 구현하는게 좋을까요? 사실 따로 Alert을 관리하는게 뭔가 모호해지는 느낌입니다.. 🤔

GREENOVER replied:
아하 저는 커스텀하게 얼럿을 구현한다면 그걸 범용적으로 많이 사용된다면
익스텐션해서 커스텀하게 구현해두고 사용하는 편입니다🙋🏻

donggeonoh committed:
style: Modify InitialVC file name and method name

donggeonoh committed:
refactor: Rename NameSpace and Divide Namespace through by role

donggeonoh committed:
refactor: Change FruitUILabel to ImageView and Add Asset
과일 이모지를 불러올 때 라벨 -> 이미지뷰로 변경
이모지는 Asset에 저장해서 불러오도록 변경
AssetNameSpace 추가
Fruit의 emoji 연산 프로퍼티 이름을 imageName으로 변경 후
AssetnameSpace에 정의된 이름을 반환하도록 구현

donggeonoh committed:
refactor: Refactor UIAlertController extension
UIAlertController extension을 지우고
UIViewController extension에 showAlert 메서드를 정의
AlertMethod 라고 하는 타입을 정의하여 확인 버튼만 있는 경우와 취소 확인이 있는 경우를
submit, suggest로 나눔

AlertNameSpace에서 notOfStock을 outOfStock으로 변경
JuiceOrderViewController 에서 호출하는 alert 메서드 수정

donggeonoh committed:
feat: Change FruitStockCollectionView scroll vertical to horizontal

GREENOVER commented on JuiceMaker/JuiceMaker/NameSpace/NameSpace.swift:
```diff
@@ -0,0 +1,36 @@
+//
+//  NameSpace.swift
```
해당 파일은 특정 네이밍을 해주지 않은 이유가 있을까요🙋🏻

donggeonoh replied:
ID나 문자열을 관리하는게 아니라 모델을 인스턴화화 할 초기값이나 레이아웃에 관련된 상수값들이라서 네이밍이 애매해서 잠시 놔뒀었는데 안바꿨었네요. 그린 주로 이런 경우엔 어떤 네이밍을 사용하시나요?

GREENOVER replied:
네임스페이스 폴더 안에 파일 만드신 네이밍처럼 명시적으로 가면 될것 같은데요..!

GREENOVER commented on JuiceMaker/JuiceMaker/Extension/UIViewController+Extension.swift:
```diff
@@ -0,0 +1,37 @@
+    ///   - submit: 확인 버튼 하나만 있는 알림창 스타일
+    ///   - suggest: 취소와 확인 버튼이 있는 알림창 스타일
+    enum AlertMethod {
+        case submit
+        case suggest
```
애매한 케이스 네이밍같아요!
얼럿이 만약 제안의 느낌이 아닌데 저런 형식으로 투버튼 형식의 얼럿을 만들고 싶다면 또 수정이 이뤄져야할까요?🤔
모듈화를 보통 어떻게 진행하는지 보면 좋을것 같아요!

donggeonoh replied:
감사합니다! 일단 얼럿이 확인버튼 하나와 확인, 취소 버튼 하나가 대부분의 경우라고 생각했는데요.
네이밍도 수정 하겠습니다! 모듈화도 공부해보겠습니다~

GREENOVER commented on JuiceMaker/JuiceMaker/View/FruitStockCollectionViewCell.swift:
```diff
@@ -0,0 +1,18 @@
+import UIKit
+
+class FruitStockCollectionViewCell: UICollectionViewCell {
+    @IBOutlet weak var fruitImageView: UIImageView!
+    @IBOutlet weak var fruitStockLabel: UILabel!
```
외부에서 알 필요가 있을까요?🙋🏻

donggeonoh replied:
access control에 신경을 좀 더 쓰도록 하겠습니다 ㅠㅠ

GREENOVER commented on JuiceMaker/JuiceMaker/View/OrderButtonCollectionViewCell.swift:
```diff
@@ -0,0 +1,16 @@
+import UIKit
+
+class OrderButtonCollectionViewCell: UICollectionViewCell {
+    @IBOutlet weak var juiceLabel: UILabel!
```
외부에서 알 필요가 있을까요?🙋🏻

donggeonoh replied:
수정 하겠습니다!

GREENOVER commented on JuiceMaker/JuiceMaker/Controller/JuiceOrderViewController.swift:
```diff
@@ -0,0 +1,192 @@
+    }
+    
+    // MARK: - Initialization
+    func initJuiceMaker() {
```
메서드명이 쥬스메이커 인스턴스를 초기화하는 느낌이 있는데 
그런 의도가 맞을까요?🙋🏻

donggeonoh replied:
네 맞습니다~ 처음 실행 시 모델을 초기화하기 위해 저런 네이밍으로 지었는데요. 의미 전달이 모호하거나 이상할까요?

GREENOVER replied:
저는 쪼오금 애매해서.. 흠 더 좋은 네이밍을 한번 고민해보죠ㅎㅎ

GREENOVER commented on JuiceMaker/JuiceMaker/Controller/JuiceOrderViewController.swift:
```diff
@@ -0,0 +1,192 @@
+    }
+    
+    // MARK: - Initialization
+    func initJuiceMaker() {
+        let storeManager = FruitStockManager(stocks: Stock.initValue)
+        let fruitStore = FruitStore(manager: storeManager)
+        
+        self.fruitStore = fruitStore
+        self.juiceMaker = JuiceMaker(store: fruitStore)
+    }
+    
+    func initCollectionView() {
```
외부에서 호출될 경우가 있을까요?🙋🏻

donggeonoh replied:
아니요 ㅠㅠ 접근 제어에 신경쓰도록 노력해야겠네요..💪

GREENOVER reviewed (approved):
고생하셨습니다!
몇가지 코멘트 남기면서 스텝3로 넘어가도록 하시죠ㅎㅎ
보완되면 좋을점들을 위주로 간결하게 남겼습니다~!
더 자세한게 있다면 스텝3에서 한번에 다루는 방향으로 하겠습니다🙋🏻
